src/common/libsdprocess/sdprocess.c

```diff
@@ -0,0 +1,1835 @@
+    char **argv;
+    char **envv;
+    int stdin_fd;
+    int stdout_fd;
+    int stderr_fd;
+
+    sd_bus *bus;
+    /* save previous bus in reconnect case */
+    sd_bus *bus_prev;
+    /* <unitname>.service */
+    char *service_name;
+    /* /org/freedesktop/systemd1/unit/<unitname>_2eservice */
+    char *service_path;
+
+    /* state watcher */
+    flux_watcher_t *w_state;
+    flux_watcher_t *w_state_prep;
+    flux_watcher_t *w_state_idle;
+    flux_watcher_t *w_state_check;
+    bool active;
+    bool active_sent;
+    bool exited;
+    bool exited_sent;
+    flux_sdprocess_state_f state_cb;
+    void *state_cb_arg;
+    uint32_t exec_main_status;
+    uint32_t exec_main_code;
+    char *active_state;
+    char *result;
+    int wait_status;
+};
+
+#define SDBUS_LOGBUF 1024
+
+static int sdprocess_state (flux_sdprocess_t *sdp,
+                            flux_sdprocess_state_f state_cb,
+                            void *arg,
+                            flux_reactor_t *reactor);
+
+static void set_and_log_errno (flux_t *h,
+                               int ret,
+                               const char *func)
+{
+    errno = -ret;
+    flux_log (h, LOG_DEBUG, "%s: %s", func, flux_strerror (errno));
+}
+
+static void set_errno_log_error (flux_t *h,
+                                 int ret,
+                                 sd_bus_error *error,
+                                 const char *func)
+{
+    assert (error);
+    errno = -ret;
+    flux_log (h, LOG_ERR, "%s: %s", func,
+              error->message ? error->message : flux_strerror (errno));
+}
+
+static void strv_destroy (char **strv)
+{
+    if (strv) {
+        char **ptr = strv;
+        while (*ptr) {
+            free (*ptr);
+            ptr++;
+        }
+        free (strv);
+    }
+}
+
+void flux_sdprocess_destroy (flux_sdprocess_t *sdp)
+{
+    if (sdp) {
+        int save_errno = errno;
+        free (sdp->unitname);
+        strv_destroy (sdp->argv);
+        strv_destroy (sdp->envv);
+
+        sd_bus_close (sdp->bus);
+        sd_bus_unref (sdp->bus);
+        sd_bus_close (sdp->bus_prev);
+        sd_bus_unref (sdp->bus_prev);
+        free (sdp->service_name);
+        free (sdp->service_path);
+
+        flux_watcher_stop (sdp->w_state);
+        flux_watcher_destroy (sdp->w_state);
+        flux_watcher_stop (sdp->w_state_prep);
+        flux_watcher_destroy (sdp->w_state_prep);
+        flux_watcher_stop (sdp->w_state_idle);
+        flux_watcher_destroy (sdp->w_state_idle);
+        flux_watcher_stop (sdp->w_state_check);
+        flux_watcher_destroy (sdp->w_state_check);
+
+        free (sdp->active_state);
+        free (sdp->result);
+        free (sdp);
+        errno = save_errno;
+    }
+}
+
+static int strv_len (char **strv)
+{
+    int len = 0;
+    if (strv) {
+        char **ptr = strv;
+        while (*ptr) {
+            len++;
+            ptr++;
+        }
+    }
+    return len;
+}
+
+static int strv_copy (char **strv, char ***cpyp)
```

Comment:
Would argz work here instead of the strv functions?

If not, maybe the strv functions should be in their own source module with a unit test?

Reply:
My understanding is that `argz` is a single buffer w/ `'\0` separating strings?  So it wouldn't work in this case.

A `strv` file sounds like a good idea, I could use it in the `sdprocess` tests as well.

Reply:
Yes but there are convenience functions for building argv style arrays of pointers and the reverse.  Not sure if it's a net win complexity to switch though - you'd probably need to consider that.

(See `argz_extract()` for example)